File: src/main/java/nl/thijsmolendijk/ibex/parse/Parser.java
package nl.thijsmolendijk.ibex.parse;

import nl.thijsmolendijk.ibex.ASTContext;
import nl.thijsmolendijk.ibex.Semantic;
import nl.thijsmolendijk.ibex.ast.Expression;
import nl.thijsmolendijk.ibex.ast.Node;
import nl.thijsmolendijk.ibex.ast.Statement;
import nl.thijsmolendijk.ibex.ast.expr.*;
import nl.thijsmolendijk.ibex.ast.stmt.*;
import nl.thijsmolendijk.ibex.scoping.Scope;
import nl.thijsmolendijk.ibex.type.FunctionType;
import nl.thijsmolendijk.ibex.type.TupleType;
import nl.thijsmolendijk.ibex.type.Type;
import nl.thijsmolendijk.ibex.util.Pair;

import java.util.ArrayList;
import java.util.List;

/**
 * Parses tokens into an AST tree.
 *
 * Created by molenzwiebel on 19-12-15.
 */
public class Parser {
    private ASTContext context;
    private Semantic sem;
    private Lexer lexer;

    private Token token;

    public Parser(String filename, String source, ASTContext context) {
        this.context = context;
        this.sem = new Semantic(context);

        this.lexer = new Lexer(filename, source);
        this.token = lexer.lex();
    }

    /**
     * @return the parsed translation unit
     */
    public TranslationUnit parseTranslationUnit() {
        TranslationUnit result = new TranslationUnit(context);
        SourceLocation start = token.getLocation();

        List<Node> contents = parseBrace();
        SourceLocation end = token.getLocation();

        sem.handleEndOfUnit(result, start, end, contents);

        return result;
    }

    /**
     * Parses a list of nodes in between '{' and '}' such that '{' node* '}'.
     */
    private List<Node> parseBrace() {
        sem.scope = new Scope(sem);
        List<Node> ret = new ArrayList<>();

        while (token.isNot(TokenType.RBRACE) && token.isNot(TokenType.EOF)) {
            switch (token.getKind()) {
                case LBRACE:
                case KW_RETURN:
                case KW_IF:
                    Statement stmt = parseStatement();
                    ret.add(stmt);
                    break;
                case KW_LET:
                    ret.add(parseVarDecl());
                    break;
                case KW_TYPE:
                    ret.add(parseTypeDecl());
                    break;
                case KW_FN:
                    ret.add(parseFnDecl());
                    break;
                default: {
                    Expression ex = parseExpr();

                    if (token.isNot(TokenType.EQUAL)) {
                        ret.add(ex);
                        break;
                    }

                    SourceLocation eqLoc = consumeToken();
                    Expression right = parseExpr();
                    if (right == null) {
                        sem.scope.end();
                        return null;
                    }

                    ret.add(new AssignExpr(ex, eqLoc, right));
                }
            }
        }

        sem.scope.end();
        return ret;
    }

    /**
     * @return the parsed var decl
     */
    public VarDecl parseVarDecl() {
        SourceLocation varLoc = consumeToken();

        String name = token.getText();
        parseToken(TokenType.IDENTIFIER, "expected identifier here", varLoc, "due to the 'let' here");

        Type givenType = null;
        if (consumeIf(TokenType.COLON)) {
            givenType = parseType();
        }

        Expression init = null;
        if (consumeIf(TokenType.EQUAL)) {
            init = parseExpr();
        }

        VarDecl result = sem.handleVarDecl(varLoc, context.getIdentifier(name), givenType, init);
        sem.addToScope(result);
        return result;
    }

    /**
     * @return the parsed type decl
     */
    public TypeDecl parseTypeDecl() {
        SourceLocation loc = consumeToken();

        Identifier name = parseIdentifier();
        parseToken(TokenType.COLON, "expected ':' in type declaration", loc, "due to the 'type' here");

        Type type = parseType();

        return sem.handleTypeDecl(loc, name, type);
    }

    /**
     * @return the parsed function decl
     */
    public FnDecl parseFnDecl() {
        SourceLocation start = consumeToken();

        Identifier name = parseIdentifier();
        if (token.isNot(TokenType.LPAREN)) {
            errorAndExit(token.getLocation(), "expected '(' in function declaration");
        }

        Type funType = parseType();
        if (!(funType instanceof FunctionType)) {
            funType = sem.handleFunctionType(funType, null, context.getUnit());
        }

        sem.scope = new Scope(sem);

        FuncExpr ex = sem.handleFunc(start, (FunctionType) funType);
        if (token.is(TokenType.LBRACE)) {
            ex.setBody(parseBraceStmt());
        }

        sem.scope.end();

        FnDecl decl = new FnDecl(name, funType, ex, start);
        sem.addToScope(decl);
        return decl;
    }

    /**
     * @return the parsed statement
     */
    public Statement parseStatement() {
        switch (token.getKind()) {
            default:
                errorAndExit(token.getLocation(), "expected statement here");
                return null;
            case KW_RETURN: return parseReturnStmt();
            case KW_IF: return parseIfStmt();
        }
    }

    /**
     * @return the parsed brace statement
     */
    public BraceStmt parseBraceStmt() {
        if (token.isNot(TokenType.LBRACE)) {
            errorAndExit(token.getLocation(), "expected '{' here");
            return null;
        }

        SourceLocation start = consumeToken();
        List<Node> res = parseBrace();

        SourceLocation end = token.getLocation();
        parseToken(TokenType.RBRACE, "expected '}' at end of brace expression", start, "to match this opening '{'");

        return new BraceStmt(start, end, res.toArray(new Node[res.size()]));
    }

    /**
     * @return the parsed return statement
     */
    public Statement parseReturnStmt() {
        SourceLocation loc = consumeToken();

        Expression val = isStartOfExpr(token) ? parseExpr() : new TupleExpr(loc, loc, new Expression[0], new Identifier[0]);
        return new ReturnStmt(loc, val);
    }

    /**
     * @return the parsed if statement
     */
    public Statement parseIfStmt() {
        SourceLocation ifLoc = consumeToken();
        Expression condition = parseExpr();

        Statement body = parseBraceStmt();

        Statement elseBody = null;
        if (consumeIf(TokenType.KW_ELSE)) {
            elseBody = token.is(TokenType.KW_IF) ? parseIfStmt() : parseBraceStmt();
        }

        return new IfStmt(ifLoc, condition, body, elseBody);
    }

    /**
     * @return the parsed expression
     */
    public Expression parseExpr() {
        List<Expression> sequence = new ArrayList<>();

        while (true) {
            Expression primary = parseExprPrimary();
            if (primary == null) {
                return null;
            }
            sequence.add(primary);

            if (!token.is(TokenType.OPERATOR)) break;
            sequence.add(parseExprOperator());
        }

        assert !sequence.isEmpty() : "Empty sequence?";
        if (sequence.size() == 1) {
            // Single element means just a simple expression, return it.
            return sequence.get(0);
        }

        return new SequenceExpr(sequence.toArray(new Expression[sequence.size()]));
    }

    /**
     * @return the parsed operator ref.
     */
    public Expression parseExprOperator() {
        SourceLocation loc = token.getLocation();
        Identifier name = parseIdentifier();
        return sem.handleIdentifier(loc, name);
    }

    /**
     * @return the parsed identifier
     */
    public Identifier parseIdentifier() {
        if (token.is(TokenType.IDENTIFIER) || token.is(TokenType.OPERATOR)) {
            Identifier res = context.getIdentifier(token.getText());
            consumeToken();
            return res;
        }

        errorAndExit(token.getLocation(), "expected identifier here");
        return null;
    }

    /**
     * @return the parsed expression
     */
    public Expression parseExprPrimary() {
        Expression result;

        switch (token.getKind()) {
            case INTEGER:
                result = sem.handleInteger(token.getLocation(), token.getText());
                consumeToken();
                break;
            case IDENTIFIER:
                result = sem.handleIdentifier(token.getLocation(), parseIdentifier());
                if (result == null) return null;
                break;
            case LPAREN:
                result = parseExprParen();
                if (result == null) return null;
                break;
            default:
                errorAndExit(token.getLocation(), "expected expression here");
                return null;
        }

        while (true) {
            SourceLocation loc = token.getLocation();

            // RESULT.field
            if (consumeIf(TokenType.DOT)) {
                parseToken(TokenType.IDENTIFIER, "expected field name here", loc, "because of this dot");

                Identifier name = context.getIdentifier(token.getText());
                result = new UnresolvedDotExpr(result, loc, token.getLocation(), name);
                consumeToken();

                continue;
            }

            // RESULT -> fun
            if (consumeIf(TokenType.ARROW)) {
                Expression fun = sem.handleIdentifier(token.getLocation(), parseIdentifier());
                result = new CallExpr(fun, result);
            }

            break;
        }

        return result;
    }

    /**
     * @return the parsed tuple expression
     */
    private Expression parseExprParen() {
        SourceLocation loc = consumeToken();
        List<Pair<Expression, Identifier>> exprs = new ArrayList<>();

        if (token.isNot(TokenType.RPAREN)) {
            do {
                Identifier fieldName = null;
                if (consumeIf(TokenType.DOT)) {
                    fieldName = parseIdentifier();
                    if (fieldName == null) return null;
                    if (!parseToken(TokenType.EQUAL, "expected '=' in tuple expression")) return null;
                }

                Expression subExpr = parseExpr();
                if (subExpr == null) return null;

                exprs.add(new Pair<>(subExpr, fieldName));
            } while (consumeIf(TokenType.COMMA));
        }

        SourceLocation rightLoc = token.getLocation();
        parseToken(TokenType.RPAREN, "expected ')' in parenthesis expression", loc, "to match this opening '('");

        return sem.handleTupleExpr(loc, exprs, rightLoc);
    }

    /**
     * @return the parsed type
     */
    public Type parseType() {
        Type result;
        switch (token.getKind()) {
            case IDENTIFIER:
                result = sem.handleTypeName(token.getLocation(), context.getIdentifier(token.getText()));
                consumeToken();
                break;
            case LPAREN:
                SourceLocation leftLoc = consumeToken();
                result = parseTupleType(leftLoc);
                parseToken(TokenType.RPAREN, "expected ')' at end of tuple type", leftLoc, "to match this opening '('");
                break;
            default:
                errorAndExit(token.getLocation(), "expected type here");
                return null;
        }

        assert result != null;

        while (true) {
            SourceLocation tokLoc = token.getLocation();
            if (consumeIf(TokenType.ARROW)) {
                Type secondHalf = parseType();
                if (secondHalf == null) return null;
                result = sem.handleFunctionType(result, tokLoc, secondHalf);

                continue;
            }

            if (consumeIf(TokenType.LBRACKET)) {
                Expression size = null;
                if (token.isNot(TokenType.RBRACKET)) {
                    size = parseExpr();
                }

                if (size != null && !(size instanceof IntegerLiteralExpr)) {
                    errorAndExit(size.getLocation(), "expected size constant here");
                }

                parseToken(TokenType.RBRACKET, "expected ']' in array type", tokLoc, "to match this opening '['");

                result = sem.handleArrayType(result, size);
                continue;
            }

            break;
        }

        return result;
    }

    /**
     * @return the parsed tuple type
     */
    public Type parseTupleType(SourceLocation leftLoc) {
        List<TupleType.TupleElement> elements = new ArrayList<>();
        if (token.isNot(TokenType.RPAREN)) {
            do {
                Identifier name = null;
                if (token.is(TokenType.IDENTIFIER) && lexer.peekToken().is(TokenType.COLON)) {
                    name = parseIdentifier();
                    parseToken(TokenType.COLON, "expected colon after identifier");
                }

                Type type = parseType();
                elements.add(new TupleType.TupleElement(name, type));
            } while (consumeIf(TokenType.COMMA));
        }

        SourceLocation rightLoc = token.getLocation();
        return sem.handleTupleType(leftLoc, elements, rightLoc);
    }

    /**
     * Same as other parseToken but without extra notes.
     */
    private boolean parseToken(TokenType expected, String errMsg) {
        return parseToken(expected, errMsg, null, null);
    }

    /**
     * If the current token is of specified type, consume it. Otherwise, print an error and stop.
     */
    private boolean parseToken(TokenType expected, String errMsg, SourceLocation noteLoc, String noteMsg) {
        if (token.is(expected)) {
            consumeToken();
            return true;
        }

        error(token.getLocation(), errMsg);
        if (noteLoc != null && noteMsg != null) {
            note(noteLoc, noteMsg);
        }

        System.err.flush();
        System.exit(1);
        return false;
    }

    /**
     * Advances the current token and returns the location of the previous one.
     */
    private SourceLocation consumeToken() {
        Token old = token;
        assert token.isNot(TokenType.EOF) : "Lexing past EOF";
        token = lexer.lex();
        return old.getLocation();
    }

    /**
     * Consumes the token if of the specified type, otherwise returns false.
     */
    private boolean consumeIf(TokenType type) {
        if (token.isNot(type)) return false;
        consumeToken();
        return true;
    }

    /**
     * @return if the specified token signals the start of an expression
     */
    private boolean isStartOfExpr(Token next) {
        return next.is(TokenType.INTEGER) || next.is(TokenType.LPAREN) || next.is(TokenType.IDENTIFIER) || next.is(TokenType.OPERATOR);
    }

    private void note(SourceLocation loc, String message) {
        Diagnostics.printMessage(Diagnostics.MessageKind.NOTE, message, loc);
    }

    private void warning(SourceLocation loc, String message) {
        Diagnostics.printMessage(Diagnostics.MessageKind.WARNING, message, loc);
    }

    private void error(SourceLocation loc, String message) {
        Diagnostics.printMessage(Diagnostics.MessageKind.ERROR, message, loc);
    }

    private void errorAndExit(SourceLocation loc, String message) {
        Diagnostics.printMessage(Diagnostics.MessageKind.ERROR, message, loc);
        System.err.flush();
        System.exit(1);
    }
}

File: src/main/kotlin/nl/thijsmolendijk/molen/parsing/Parser.kt
package nl.thijsmolendijk.molen.parsing

import nl.thijsmolendijk.molen.ast.*
import nl.thijsmolendijk.molen.parsing.TokenType.*
import nl.thijsmolendijk.molen.type.FunctionType
import nl.thijsmolendijk.molen.type.TupleElement
import nl.thijsmolendijk.molen.type.TupleType
import nl.thijsmolendijk.molen.type.Type
import java.util.*

/**
 * Parses source file into AST.
 *
 * Created by molenzwiebel on 09-01-16.
 */
fun main(args: Array<String>) {
    val i = Scanner(System.`in`)
    while (true) {
        val p = Parser("<repl>", i.nextLine())
        val q = p.parseTranslationUnit()
        println(q.text)
    }
}

class Parser(val filename: String, val source: String) {
    private val lexer = Lexer(filename, source)
    private var token = lexer.lex()

    public fun parseTranslationUnit(): TranslationUnit {
        return TranslationUnit(token.location, parseList())
    }

    private fun parseList(): List<Node> {
        val ret: MutableList<Node> = arrayListOf()

        while (token isnt RBRACE && token isnt EOF) {
            ret.add(when (token.kind) {
                LBRACE, KW_RETURN, KW_IF, KW_FOR -> parseStatement()
                KW_LET, KW_VAL, KW_VAR -> parseVarDecl()
                KW_TYPEALIAS -> parseTypeDecl()
                KW_FN -> parseFunDecl()
                KW_IMPORT -> parseImportDecl()
                else -> parseExpr()
            })
        }

        return ret
    }

    private fun parseBraceStmt(): BraceStmt {
        val begin = parseToken(LBRACE, "expected '{' here").location
        val res = BraceStmt(begin, parseList())
        parseToken(RBRACE, "expected '}' here", begin, "to match the '{' here")
        return res
    }

    private fun parseNode(): Node {
        return when (token.kind) {
            KW_VAL, KW_VAR, KW_LET, KW_IF, KW_FOR, KW_RETURN -> parseStatement()
            else -> parseExpr()
        }
    }

    private fun parseStatement(): Statement {
        return when (token.kind) {
            KW_VAL, KW_VAR, KW_LET -> parseVarDecl()
            KW_IF -> parseIf()
            KW_FOR -> parseFor()
            KW_RETURN -> parseReturn()
            else -> {
                abrt("expected statement here, received ${token.kind}")
                throw RuntimeException("Unreachable")
            }
        }
    }

    private fun parseIf(): Statement {
        val start = consumeToken(KW_IF)
        val cond = parseExpr()
        val body = parseBraceStmt()
        var elseBody: BraceStmt? = null

        consumeIf(KW_ELSE) {
            elseBody = if (token kind KW_IF) BraceStmt(it, listOf(parseIf())) else parseBraceStmt()
        }

        return IfStmt(start, cond, body, elseBody)
    }

    private fun parseFor(): Statement {
        val start = consumeToken(KW_FOR)

        val init = if (consumeIf(COMMA)) null else { val ret = parseNode(); parseToken(COMMA, "expected ',' after for initialization", start, "due to the 'for' here"); ret }
        val cond = parseExpr()
        parseToken(COMMA, "expected ',' after for condition", start, "due to the 'for' here")
        val step = if (token kind LBRACE) null else parseNode()

        val body = parseBraceStmt()
        return ForStmt(start, init, cond, step, body)
    }

    private fun parseReturn(): Statement {
        return ReturnStmt(consumeToken(KW_RETURN), parseExpr())
    }

    private fun parseExpr(): Expression {
        val els = arrayListOf<Expression>()

        while (true) {
            val expr = parseExprPrimary()
            els.add(expr)

            if (token isnt OPERATOR) break
            val op = parseToken(OPERATOR, "expected operator after expression")
            els.add(UnresolvedDeclRef(op.location, op.text))
        }

        return if (els.size == 1) els.first() else SequenceExpr(els.first().location, els)
    }

    private fun parseExprPrimary(): Expression {
        val tok = token

        var ret = when (token.kind) {
            KW_TRUE, KW_FALSE -> {
                consumeToken()
                BooleanLiteral(tok.location, tok.text == "true")
            }

            LPAREN -> parseTupleLiteral()

            INTEGER -> {
                consumeToken(INTEGER)
                IntLiteral(tok.location, tok.text.toInt())
            }

            IDENTIFIER -> {
                consumeToken(IDENTIFIER)
                UnresolvedDeclRef(tok.location, tok.text)
            }

            else -> {
                abrt("expected expression here, received ${token.kind}")
                throw RuntimeException("Unreachable")
            }
        }

        // Handle suffixes
        while (true) {
            // ret.foo
            if (consumeIf(COLON)) {
                ret = UnresolvedDotExpr(tok.location, ret, parseToken(IDENTIFIER, "expected identifier after '.'").text)
                continue
            }

            // ret(foo)
            if (token kind LPAREN) {
                ret = CallExpr(tok.location, ret, parseTupleLiteral())
                continue
            }

            // ret = foo
            if (consumeIf(EQUAL)) {
                ret = AssignExpr(tok.location, ret, parseExpr())
                continue
            }

            // FIXME: Array access

            break
        }

        return ret
    }

    private fun parseTupleLiteral(): TupleLiteral {
        val begin = consumeToken(LPAREN)
        val entries = arrayListOf<TupleLiteralElement>()

        if (token isnt RPAREN) {
            do {
                val name = if (consumeIf(COLON)) {
                    val name = parseToken(IDENTIFIER, "expected identifier after '.' in tuple literal").text
                    parseToken(EQUAL, "expected '=' after .identifier in tuple literal")
                    name
                } else null

                entries.add(TupleLiteralElement(name, parseExpr()))
            } while (consumeIf(COMMA))
        }

        parseToken(RPAREN, "expected ')' here", begin, "to match the '(' here")
        return TupleLiteral(begin, entries)
    }

    private fun parseVarDecl(): VarDecl {
        val loc = consumeToken()
        val name = parseToken(IDENTIFIER, "expected identifier here", loc, "due to the 'let'/'var'/'val' here").text

        var givenType: Type? = null
        consumeIf(COLON) {
            givenType = parseType()
        }

        var init: Expression? = null
        consumeIf(EQUAL) {
            init = parseExpr()
        }

        return VarDecl(loc, name, givenType, init)
    }

    private fun parseTypeDecl(): TypeAliasDecl {
        val loc = consumeToken(KW_TYPEALIAS)
        val name = parseToken(IDENTIFIER, "expected identifier in typealias declaration", loc, "due to the 'typealias' here").text
        parseToken(COLON, "expected ':' in typealias declaration", loc, "due to the 'typealias' here")

        return TypeAliasDecl(loc, name, parseType())
    }

    private fun parseFunDecl(): FunDecl {
        val start = consumeToken(KW_FN)
        val name = parseToken(IDENTIFIER, "expected identifier after 'fun'", start, "due to the 'fun' here").text

        if (token isnt LPAREN) abrt("expected '(' in function declaration")
        var funType = parseType()
        if (funType !is FunctionType) funType = FunctionType(funType, TupleType(arrayOf()))

        val argDecls = arrayListOf<ArgDecl>()
        for (arg in (funType.args as TupleType).elements) {
            argDecls.add(ArgDecl(start, arg.name ?: throw RuntimeException("Function argument must be named"), arg.type))
        }

        return FunDecl(start, name, funType, ClosureExpr(start, argDecls, BraceStmt(start, parseList())))
    }

    private fun parseImportDecl(): ImportDecl {
        val loc = consumeToken(KW_IMPORT)
        val lit = parseExpr()
        if (lit !is StringLiteral) abrt("expected string literal after 'import'", lit.location)
        return ImportDecl(loc, (lit as StringLiteral).value)
    }

     fun parseType(): Type {
        var result = when (token.kind) {
            IDENTIFIER -> {
                //FIXME: Perform actual lookup
                val decl = TypeAliasDecl(token.location, token.text, null)
                consumeToken()
                decl.aliasType
            }
            LPAREN -> {
                val leftLoc = consumeToken(LPAREN)
                val res = parseTupleType()
                parseToken(RPAREN, "expected ')' at end of tuple type", leftLoc, "to match this opening '('")
                res
            }
            else -> {
                abrt("expected type here")
                throw RuntimeException("Unreachable")
            }
        }

        while (true) {
            val loc = token.location

            consumeIf(ARROW) {
                val secondHalf = parseType()
                result = FunctionType(result, secondHalf)
            }

            if (token.location == loc) break
        }

        return result
    }

    private fun parseTupleType(): Type {
        val elements = arrayListOf<TupleElement>()
        if (token isnt RPAREN) {
            do {
                var name: String? = if (token kind IDENTIFIER && lexer.peek() kind COLON) {
                    val name = token.text
                    consumeToken()
                    parseToken(COLON, "expected colon after tuple element name")
                    name
                } else null

                elements.add(TupleElement(name, parseType()))
            } while (consumeIf(COMMA))
        }

        return TupleType(elements.toTypedArray())
    }

    private fun consumeToken(ty: TokenType = UNKNOWN): SourceLocation {
        if (ty != UNKNOWN) {
            if (token isnt ty) throw IllegalStateException("Consuming token of wrong kind. $ty expected, ${token.kind} given")
        }
        val old = token
        token = lexer.lex();
        return old.location
    }

    private fun consumeIf(type: TokenType): Boolean {
        if (token isnt type) return false
        consumeToken()
        return true
    }

    private fun <R> consumeIf(type: TokenType, fn: (SourceLocation) -> R): R? {
        if (token isnt type) return null
        val loc = consumeToken();
        return fn(loc)
    }

    private fun parseToken(expected: TokenType, msg: String, noteLoc: SourceLocation? = null, noteMsg: String? = null): Token {
        if (token kind expected) {
            val tok = token
            consumeToken()
            return tok
        }

        Diagnostics.printMessage(Diagnostics.MessageKind.ERROR, msg, token.location)
        if (noteLoc != null && noteMsg != null) {
            Diagnostics.printMessage(Diagnostics.MessageKind.NOTE, noteMsg, noteLoc)
        }

        val ex = Exception()
        ex.printStackTrace()

        System.err.flush()
        System.out.flush()
        System.exit(1)

        throw RuntimeException("Unreachable")
    }

    private fun abrt(msg: String, loc: SourceLocation = token.location) {
        Diagnostics.printMessage(Diagnostics.MessageKind.ERROR, msg, loc)
        System.exit(1)
    }
}